src/mscorlib/src/System/ArraySegment.cs

```diff
@@ -110,6 +110,87 @@ public int Count
             }
         }
 
+        public int Length => Count;
+
+        public T this[int index]
+        {
+            get
+            {
+                if (_array == null)
+                    ThrowHelper.ThrowInvalidOperationException(ExceptionResource.InvalidOperation_NullArray);
+                if (index < 0 || index >= _count)
+                    ThrowHelper.ThrowArgumentOutOfRange_IndexException();
+                Contract.EndContractBlock();
+
+                return _array[_offset + index];
+            }
+
+            set
+            {
+                if (_array == null)
+                    ThrowHelper.ThrowInvalidOperationException(ExceptionResource.InvalidOperation_NullArray);
+                if (index < 0 || index >= _count)
+                    ThrowHelper.ThrowArgumentOutOfRange_IndexException();
+                Contract.EndContractBlock();
+
+                _array[_offset + index] = value;
```

Comment:
@AlexGhiondea
1. segment `_offset + _count` can not be more that `array.count` it is checked at segment constructor.
2. We check that `index >= 0 and index < _count` here so it can not be more segment _count and can not be more array count because segment `_offset + _count` less than array count. See step 1.